File: prototype/vnext/agentic_vnext/framework_lock.py
"""Create and validate the technical inputs pinned by a Framework lock."""

from __future__ import annotations

from copy import deepcopy
from dataclasses import dataclass
from pathlib import Path
import re
from typing import Any

import yaml

from .model import RuleIndex, canonical_digest
from .schema import SCHEMA_BUNDLE_VERSION, default_schema_registry
from .versions import (
    APPLICATION_PROTOCOL_VERSION,
    CANONICALIZATION_VERSION,
    CONTEXT_COMPILER_VERSION,
    DATA_MODEL_VERSION,
    DETECTOR_ID,
    DETECTOR_VERSION,
    EXPLANATION_VERSION,
    FRAMEWORK_LOCK_SCHEMA_VERSION,
    FRAMEWORK_RELEASE,
    KERNEL_VERSION,
    PROJECT_SNAPSHOT_PROTOCOL_VERSION,
    RULE_COMPILER_VERSION,
    SIGNED_FRAMEWORK_LOCK_SCHEMA_VERSION,
)


@dataclass(frozen=True)
class FrameworkLock:
    """Validated immutable view of the runtime and Rule set identity."""

    manifest: dict[str, Any]
    digest: str

    def as_dict(self) -> dict[str, Any]:
        return deepcopy(self.manifest)


def build_framework_lock(
    rule_source: dict[str, Any],
    rule_index: RuleIndex,
) -> dict[str, Any]:
    """Build the exact lock expected by this runtime.

    Generation is a technical operation: it records versions and content digests
    but never decides whether a Rule change is semantically acceptable.
    """

    return {
        "schema_version": FRAMEWORK_LOCK_SCHEMA_VERSION,
        "framework_release": FRAMEWORK_RELEASE,
        "protocols": {
            "application": APPLICATION_PROTOCOL_VERSION,
            "canonicalization": CANONICALIZATION_VERSION,
            "context_compiler": CONTEXT_COMPILER_VERSION,
            "data_model": DATA_MODEL_VERSION,
            "explanation": EXPLANATION_VERSION,
            "kernel": KERNEL_VERSION,
            "project_snapshot": PROJECT_SNAPSHOT_PROTOCOL_VERSION,
            "rule_compiler": RULE_COMPILER_VERSION,
        },
        "detectors": {
            DETECTOR_ID: DETECTOR_VERSION,
        },
        "schema_bundle": {
            "version": SCHEMA_BUNDLE_VERSION,
            "digest": default_schema_registry().digest,
        },
        "rule_set": {
            "source_digest": canonical_digest(rule_source),
            "index_digest": rule_index.digest,
        },
    }


def validate_framework_lock(
    lock_source: dict[str, Any],
    rule_source: dict[str, Any],
    rule_index: RuleIndex,
) -> FrameworkLock:
    """Reject any runtime or Rule input that differs from the pinned manifest."""

    expected = build_framework_lock(rule_source, rule_index)
    differences = _differences(expected, _comparable_core_lock(lock_source))
    if differences:
        raise ValueError(
            "framework lock mismatch:\n- " + "\n- ".join(differences)
        )
    manifest = deepcopy(lock_source)
    return FrameworkLock(
        manifest=manifest,
        digest=canonical_digest(manifest),
    )


def _comparable_core_lock(lock_source: dict[str, Any]) -> dict[str, Any]:
    """Remove only the delivery-owned v2 extension before runtime comparison."""

    version = lock_source.get("schema_version")
    if version == FRAMEWORK_LOCK_SCHEMA_VERSION:
        return deepcopy(lock_source)
    if version != SIGNED_FRAMEWORK_LOCK_SCHEMA_VERSION:
        raise ValueError(f"unsupported framework lock Schema: {version!r}")
    artifact = lock_source.get("release_artifact")
    expected_fields = {"artifact_digest", "source_id", "signer_key_id"}
    if not isinstance(artifact, dict) or set(artifact) != expected_fields:
        raise ValueError(
            "framework lock v2 release_artifact fields must be exactly "
            f"{sorted(expected_fields)!r}"
        )
    for field in expected_fields:
        if not isinstance(artifact[field], str) or not artifact[field]:
            raise ValueError(
                f"framework lock release_artifact.{field} "
                "must be a non-empty string"
            )
    if re.fullmatch(r"sha256:[0-9a-fA-F]{64}", artifact["artifact_digest"]) is None:
        raise ValueError(
            "framework lock release_artifact.artifact_digest "
            "must be a SHA-256 digest"
        )
    comparable = deepcopy(lock_source)
    comparable["schema_version"] = FRAMEWORK_LOCK_SCHEMA_VERSION
    del comparable["release_artifact"]
    return comparable


def load_framework_lock(path: str | Path) -> dict[str, Any]:
    with Path(path).open(encoding="utf-8") as stream:
        value = yaml.safe_load(stream)
    if not isinstance(value, dict):
        raise ValueError("framework lock must be a mapping")
    return value


def _differences(
    expected: Any,
    actual: Any,
    path: str = "",
) -> list[str]:
    """Return stable, field-level differences for actionable startup errors."""

    if isinstance(expected, dict) and isinstance(actual, dict):
        differences: list[str] = []
        for key in sorted(set(expected) | set(actual)):
            child_path = f"{path}.{key}" if path else str(key)
            if key not in expected:
                differences.append(f"{child_path}: unexpected field")
            elif key not in actual:
                differences.append(f"{child_path}: missing field")
            else:
                differences.extend(
                    _differences(expected[key], actual[key], child_path)
                )
        return differences
    if expected != actual:
        return [f"{path}: expected {expected!r}, got {actual!r}"]
    return []

File: src/framework_lock.rs
//! Build and validate the complete technical identity of one Framework runtime.
//!
//! Lock generation records versions and content digests. It does not decide
//! whether a Rule or Schema change is semantically acceptable.

use crate::canonical_digest;
use crate::context::CONTEXT_COMPILER_VERSION;
use crate::detection::{DETECTOR_ID, DETECTOR_VERSION};
use crate::kernel::KERNEL_VERSION;
use crate::project::PROJECT_SNAPSHOT_PROTOCOL_VERSION;
use crate::rules::{RULE_COMPILER_VERSION, RuleIndex};
use crate::schema::{SCHEMA_BUNDLE_VERSION, SchemaRegistry};
use crate::{APPLICATION_PROTOCOL_VERSION, CANONICALIZATION_VERSION};
use serde_json::{Value, json};
use std::fmt;

pub const DATA_MODEL_VERSION: &str = "3";
pub const EXPLANATION_VERSION: &str = "2";
pub const FRAMEWORK_LOCK_SCHEMA_VERSION: &str = "1";
pub const SIGNED_FRAMEWORK_LOCK_SCHEMA_VERSION: &str = "2";
pub const FRAMEWORK_RELEASE: &str = "adf-dev";

#[derive(Debug, Clone, PartialEq)]
pub struct FrameworkLock {
    pub manifest: Value,
    pub digest: String,
}

pub fn build_framework_lock(
    rule_source: &Value,
    rule_index: &RuleIndex,
    schema_registry: &SchemaRegistry,
) -> Result<Value, FrameworkLockError> {
    let source_digest = canonical_digest(rule_source)
        .map_err(|error| FrameworkLockError::new(error.to_string()))?;
    Ok(json!({
        "schema_version": FRAMEWORK_LOCK_SCHEMA_VERSION,
        "framework_release": FRAMEWORK_RELEASE,
        "protocols": {
            "application": APPLICATION_PROTOCOL_VERSION,
            "canonicalization": CANONICALIZATION_VERSION,
            "context_compiler": CONTEXT_COMPILER_VERSION,
            "data_model": DATA_MODEL_VERSION,
            "explanation": EXPLANATION_VERSION,
            "kernel": KERNEL_VERSION,
            "project_snapshot": PROJECT_SNAPSHOT_PROTOCOL_VERSION,
            "rule_compiler": RULE_COMPILER_VERSION,
        },
        "detectors": {
            DETECTOR_ID: DETECTOR_VERSION,
        },
        "schema_bundle": {
            "version": SCHEMA_BUNDLE_VERSION,
            "digest": schema_registry.digest(),
        },
        "rule_set": {
            "source_digest": source_digest,
            "index_digest": rule_index.digest,
        },
    }))
}

pub fn validate_framework_lock(
    lock_source: &Value,
    rule_source: &Value,
    rule_index: &RuleIndex,
    schema_registry: &SchemaRegistry,
) -> Result<FrameworkLock, FrameworkLockError> {
    let expected = build_framework_lock(rule_source, rule_index, schema_registry)?;
    let comparable = comparable_core_lock(lock_source)?;
    let differences = differences(&expected, &comparable, "");
    if !differences.is_empty() {
        return Err(FrameworkLockError::new(format!(
            "framework lock mismatch:\n- {}",
            differences.join("\n- ")
        )));
    }
    let digest = canonical_digest(lock_source)
        .map_err(|error| FrameworkLockError::new(error.to_string()))?;
    Ok(FrameworkLock {
        manifest: lock_source.clone(),
        digest,
    })
}

/// Normalize a signed-delivery v2 lock to the v1 runtime identity.
///
/// Delivery owns signature/source verification. The Application still compares
/// every runtime, Rule, and Schema field exactly, without coupling itself to a
/// transport implementation.
fn comparable_core_lock(lock_source: &Value) -> Result<Value, FrameworkLockError> {
    match lock_source.get("schema_version").and_then(Value::as_str) {
        Some(FRAMEWORK_LOCK_SCHEMA_VERSION) => Ok(lock_source.clone()),
        Some(SIGNED_FRAMEWORK_LOCK_SCHEMA_VERSION) => {
            validate_release_artifact_shape(lock_source)?;
            let mut comparable = lock_source.clone();
            let object = comparable
                .as_object_mut()
                .ok_or_else(|| FrameworkLockError::new("framework lock must be a mapping"))?;
            object.insert(
                "schema_version".to_owned(),
                Value::String(FRAMEWORK_LOCK_SCHEMA_VERSION.to_owned()),
            );
            object.remove("release_artifact");
            Ok(comparable)
        }
        Some(version) => Err(FrameworkLockError::new(format!(
            "unsupported framework lock Schema: {version}"
        ))),
        None => Err(FrameworkLockError::new(
            "framework lock schema_version must be a string",
        )),
    }
}

fn validate_release_artifact_shape(lock_source: &Value) -> Result<(), FrameworkLockError> {
    let artifact = lock_source
        .get("release_artifact")
        .and_then(Value::as_object)
        .ok_or_else(|| {
            FrameworkLockError::new("framework lock v2 release_artifact must be a mapping")
        })?;
    let mut actual = artifact.keys().map(String::as_str).collect::<Vec<_>>();
    actual.sort_unstable();
    let expected = ["artifact_digest", "signer_key_id", "source_id"];
    if actual != expected {
        return Err(FrameworkLockError::new(format!(
            "framework lock v2 release_artifact fields must be exactly {expected:?}"
        )));
    }
    for field in expected {
        let value = artifact
            .get(field)
            .and_then(Value::as_str)
            .filter(|value| !value.is_empty())
            .ok_or_else(|| {
                FrameworkLockError::new(format!(
                    "framework lock release_artifact.{field} must be a non-empty string"
                ))
            })?;
        if field == "artifact_digest"
            && !(value.starts_with("sha256:")
                && value.len() == "sha256:".len() + 64
                && value["sha256:".len()..]
                    .bytes()
                    .all(|byte| byte.is_ascii_hexdigit()))
        {
            return Err(FrameworkLockError::new(
                "framework lock release_artifact.artifact_digest must be a SHA-256 digest",
            ));
        }
    }
    Ok(())
}

fn differences(expected: &Value, actual: &Value, path: &str) -> Vec<String> {
    if let (Some(expected), Some(actual)) = (expected.as_object(), actual.as_object()) {
        let mut keys = expected.keys().chain(actual.keys()).collect::<Vec<_>>();
        keys.sort();
        keys.dedup();
        let mut output = Vec::new();
        for key in keys {
            let child_path = if path.is_empty() {
                key.clone()
            } else {
                format!("{path}.{key}")
            };
            match (expected.get(key), actual.get(key)) {
                (None, Some(_)) => output.push(format!("{child_path}: unexpected field")),
                (Some(_), None) => output.push(format!("{child_path}: missing field")),
                (Some(expected), Some(actual)) => {
                    output.extend(differences(expected, actual, &child_path));
                }
                (None, None) => unreachable!("key came from one of the objects"),
            }
        }
        return output;
    }
    if expected == actual {
        Vec::new()
    } else {
        vec![format!("{path}: expected {expected}, got {actual}")]
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FrameworkLockError {
    message: String,
}

impl FrameworkLockError {
    fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for FrameworkLockError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl std::error::Error for FrameworkLockError {}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn reports_nested_missing_and_unexpected_fields_in_order() {
        assert_eq!(
            differences(
                &json!({"a": {"required": "1"}}),
                &json!({"a": {"unexpected": "1"}}),
                "",
            ),
            vec![
                "a.required: missing field",
                "a.unexpected: unexpected field",
            ]
        );
    }
}
